src/NuGet.Core/NuGet.Commands/ListCommand/ListCommandRunner.cs

```diff
@@ -0,0 +1,172 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Commands.ListCommand;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.Commands
+{
+    /// <summary>
+    /// Command Runner used to run the business logic for nuget list command
+    /// </summary>
+    public class ListCommandRunner : IListCommandRunner
+    {
+        /// <summary>
+        /// Executes the logic for nuget list command.
+        /// </summary>
+        /// <returns></returns>
+        public async Task ExecuteCommand(ListArgs listArgs)
+
+        {
+            //Create SourceFeed for each packageSource
+            var sourceFeeds = new List<ListResource>();
+            // this is to avoid duplicate remote calls in case of duplicate final endpoints (Ex. api/index.json and /api/v2/ point to the same target)
+            var sources = new HashSet<string>();
+
+            foreach (PackageSource packageSource in listArgs.ListEndpoints)
+            {
+                var sourceRepository = Repository.Factory.GetCoreV3(packageSource.Source);
+                var feed = await sourceRepository.GetResourceAsync<ListResource>(listArgs.CancellationToken);
+                
+                if (feed != null)
+                {
+                    if (sources.Add(feed.Source))
+                    {
+                        sourceFeeds.Add(feed);
+                    }
+                }
+                else
+                {
+                    listArgs.Logger.LogWarning(string.Format(listArgs.ListCommandListNotSupported,packageSource.Source));
+                }
+            }
+
+            var allPackages = new List<IEnumerableAsync<IPackageSearchMetadata>>();
+            var log = listArgs.IsDetailed ? listArgs.Logger : NullLogger.Instance;
+            foreach (var feed in sourceFeeds)
+            {
+                    var packagesFromSource =
+                        await feed.ListAsync(listArgs.Arguments.FirstOrDefault(), listArgs.Prerelease, listArgs.AllVersions,
+                            listArgs.IncludeDelisted, log, listArgs.CancellationToken);
+                    allPackages.Add(packagesFromSource); 
+            }
+            CompareIPackageSearchMetadata comparer = new CompareIPackageSearchMetadata();
+            await PrintPackages(listArgs, new AggregateEnumerableAsync<IPackageSearchMetadata>(allPackages, comparer, comparer).GetEnumeratorAsync());
+        }
+
+        private class CompareIPackageSearchMetadata : IComparer<IPackageSearchMetadata>, IEqualityComparer<IPackageSearchMetadata>
+        {
+            private readonly PackageIdentityComparer _comparer;
+            public CompareIPackageSearchMetadata()
```

Comment:
>CompareIPackageSearchMetadata [](start = 19, length = 29)

Nit: this can be done with a field initializer so you don't have to declare the ctor